web/firefoxcom.js

```diff
@@ -198,6 +198,27 @@ class MozL10n {
   }
 })();
 
+(function listenZoomEvents() {
+  const events = [
+    'zoomin',
+    'zoomout',
+    'zoomreset',
+  ];
+  let handleEvent = function({ type, detail, }) {
+    if (!PDFViewerApplication.initialized) {
+      return;
+    }
+    PDFViewerApplication.eventBus.dispatch(type, {
+      source: window,
+      ignoreDuplicate: (type === 'zoomreset' ? true : undefined),
```

Comment:
With the complete patch, see https://github.com/mozilla/pdf.js/pull/10652#issuecomment-474779393, there's now three different places a "reset zoom to auto" signal can come from:
 - With focus in the location bar (or menu), pressing <kbd>Ctrl</kbd>+<kbd>0</kbd> will dispatch a `PDFJS:ZoomReset` browser event which is then forwarded to the viewer as a `zoomreset` window event.
 - Clicking the "Reset zoom level" button in the Firefox menu, which again goes through the `PDFJS:ZoomReset` to `zoomreset` event conversion to reach the viewer.
 - With focus in the PDF.js viewer, pressing <kbd>Ctrl</kbd>+<kbd>0</kbd> which is handled in https://github.com/mozilla/pdf.js/blob/844aecf9e3bcec7c8e4fb3f35ebadb562eec8013/web/app.js#L2188-L2199

   Note how we're *purposely* not `preventDefault`ing in this case, to ensure that the browser zoom level (if it's non-default) is reset *first* before applying the PDF.js resetting.
However, with the changes I'm proposing for `mozilla-central` this case will now *also* send dispatch a `PDFJS:ZoomReset` browser event. Hence using <kbd>Ctrl</kbd>+<kbd>0</kbd> when the viewer has focus will thus result in the following things ordering:
   1. <kbd>Ctrl</kbd>+<kbd>0</kbd> is pressed, which is allowed to pass from the viewer to the browser.
   2. The `setTimeout` finishes, which causes the PDF.js zoom level to be reset via `PDFViewerApplication.zoomReset`.
   3. The asynchronous browser `PDFJS:ZoomReset` event is handled, in `PdfStreamConverter.jsm`, and forwarded to the viewer as a window `zoomreset` event.
   4. The `zoomreset` event is now forwarded on the `eventBus`, thus leading to one more (almost immediate) `PDFViewerApplication.zoomReset` invocation as a result of the original <kbd>Ctrl</kbd>+<kbd>0</kbd> press.

Since we must, in order to always be able to reset the PDF.js viewer zoom level correctly, dispatch a browser `PDFJS:ZoomReset` event this construction was simply my attempt at ignoring the double `PDFViewerApplication.zoomReset` every single time that <kbd>Ctrl</kbd>+<kbd>0</kbd> is used in the viewer. Hopefully this helps clear things up :-)

---

> I'm fine with merging this if there is also an upstream reviewer. Perhaps @brendandahl can take a look at the Bugzilla patch, or find someone upstream that can?

According to his Bugzilla signature, Brendan is currently out of the office: "Away until Mar 29th".

I'm hoping that this patch can land here first though, since you shouldn't generally edit the PDF.js code directly in mozilla-central anyway (I just needed to have a complete patch to actually test it).
Furthermore, the PDF.js viewer parts of this puzzle should be fairly well separated from the actual browser changes needed as well.